src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1895 @@
+                                _exit(EXIT_FAILURE);
+                        }
+
+                        log_debug("Spawning sysupdate: %s", s);
+                }
+
+                r = invoke_callout_binary(sysupdate_binary_path(), args);
+                log_error_errno(r, "Failed to execute systemd-sysupdate: %m");
+                _exit(EXIT_FAILURE);
+        }
+
+        pipe[1] = safe_close(pipe[1]);
+        f = take_fdopen(&pipe[0], "r");
+        if (!f)
+                return -errno;
+
+        r = sd_json_parse_file(f, "stdout", 0, &v, NULL, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse JSON: %m");
+
+        *ret = TAKE_PTR(v);
+        return 0;
+}
+
+static int target_ensure_properties(Target *t) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *v = NULL;
+        _cleanup_free_ char *target_arg = NULL;
+        int r;
+
+        assert(t);
+
+        if (t->version && t->appstream)
+                return 0;
+
+        r = target_get_argument(t, &target_arg);
+        if (r < 0)
+                return r;
+
+        r = sysupdate_run_simple(&v, "--offline", "list", target_arg, NULL);
+        if (r < 0)
+                return r;
+
+        r = cache_string_from_json(&t->version, "current", v);
+        if (r < 0)
+                return r;
+
+        r = cache_strv_from_json(&t->appstream, "appstream-url", v);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+static BUS_DEFINE_PROPERTY_GET_ENUM(target_property_get_class, target_class, TargetClass);
+
+static int target_method_list_finish(
+                sd_bus_message *msg,
+                const Job *j,
+                sd_json_variant *json,
+                sd_bus_error *error) {
+
+        sd_json_variant *v;
+        _cleanup_strv_free_ char **versions = NULL;
+        _cleanup_(sd_bus_message_unrefp) sd_bus_message *reply = NULL;
+        int r;
+
+        assert(json);
+
+        /* Try our best since we have the data anyway, but isn't fatal if it fails */
+        (void) cache_string_from_json(&j->target->version, "current", json);
+        (void) cache_strv_from_json(&j->target->appstream, "appstream-url", json);
+
+        v = sd_json_variant_by_key(json, "all");
+        if (!v)
+                return -EINVAL;
+
+        r = sd_json_variant_strv(v, &versions);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_new_method_return(msg, &reply);
+        if (r < 0)
+                return r;
+
+        r = sd_bus_message_append_strv(reply, versions);
+        if (r < 0)
+                return r;
+
+        return sd_bus_send(NULL, reply, NULL);
+}
+
+static int target_method_list(sd_bus_message *msg, void *userdata, sd_bus_error *error) {
+        Target *t = ASSERT_PTR(userdata);
+        _cleanup_(job_freep) Job *j = NULL;
+        bool offline;
+        int r;
+
+        assert(msg);
+
+        r = sd_bus_message_read(msg, "b", &offline);
```

Comment:
when we wrote sd-bus we still cared about C90 and didn't use "bool" in public APIs. Hence the dbus boolean type maps to "int", not "bool". And those two types typically have different sizes, hence this is a memory corruption. i.e. change "bool offline" to "int offline";

Reply:
(internally, please use C99 "bool" wherever you can, but when you get booleans you need to read them into int first)